examples/iguana-example-photon-ML.cc

```diff
@@ -0,0 +1,73 @@
+/// @begin_doc_example
```

Comment:
Ahh, so sometimes you may actually prefer
```cpp
for(int row = 0; row < some_bank.getRows(); row++) // does not care about upstream filters
```
instead of
```cpp
for(const auto& row : some_bank.getRowList()) // uses upstream filters
```
Can you use the `getRows()` version where necessary, or could you clarify in the documentation, if indeed this algorithm needs to come before others? If you do need `getRows()`, it would help to put a comment nearby saying something like "we need to loop over ALL the rows here".

Reply:
Yes, in this case, I will be using the aforementioned `int row` for loop method to automatically set this. I'll run a test now with it.

Reply:
@c-dilks in this line of `PhotonGBTFilter::Run` 

```
    particleBank.getMutableRowList().filter([this, &caloBank, &calo_map, runnum](auto bank, auto row) {
        auto pid = bank.getInt("pid", row);
        if (pid != 22) return true;
        return Filter(bank, caloBank, calo_map, row, runnum);
    });
```

what would be the correct syntax for looping over all rows, and not those already filtered?

Reply:
No, `getMutableRowList()` gets the filtered list of rows. Perhaps you could do something like
```cpp
hipo::bank::list_t rowlist_new = {};
auto rowlist_old = particleBank.getRowList();
for(int row = 0; row < particleBank.getRows(); row++) { // NOTE: loop over ALL particles, not just those that survive upstream filters
  auto pid = bank.getInt("pid", row);
  
  bool was_not_filtered = std::find(rowlist_old.begin(), rowlist_old.end(), row) != rowlist_old.end();

  if(pid != 22) {
    if(was_not_filtered)
      rowlist_new.push_back(row);
  }

  else {
    auto result = Filter(bank, caloBank, calo_map, row, runnum);
    if(was_not_filtered && result)
      filter_list.push_back(row);
  }
}
particleBank.getMutableRowList().setList(rowlist_new);
```
I'm not so sure what you want to do here exactly, but hopefully this example usage helps. For more info:
- `hipo::bank::rowlist`: https://github.com/gavalian/hipo/blob/7a4308f22241ca1b6c599d4ac689296dc3724feb/hipo4/bank.h#L213
- examples: https://github.com/gavalian/hipo/blob/master/examples/bankRowList.cc

Reply:
Ah, in fact, I don't think it would be necessary to do this. We should be able to just keep the `PhotonGBTFilter::Run` code as 
```
    particleBank.getMutableRowList().filter([this, &caloBank, &calo_map, runnum](auto bank, auto row) {
        auto pid = bank.getInt("pid", row);
        if (pid != 22) return true;
        return Filter(bank, caloBank, calo_map, row, runnum);
    });
 ```
 This is because we are still looping over the unfiltered particleBank in `PhotonGBTFilter::Filter` 
 ```
 for(int inner_row = 0; inner_row < particleBank.getRows(); inner_row++) {
 ```
 It is therefore fine to keep `PhotonGBTFilter::Run` unchanged. I do not see a difference in the validator plots when swapping the order of EventBuilderFilter and PhotonGBTFilter algorithms.